File: array/valid-sudoku.py
# https://leetcode.com/problems/valid-sudoku/
# tags - array
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        def areDuplicates(arr):
            return len(arr) > len(set(arr))
        
        def excludeEmpty(arr):
            return [x for x in arr if x != '.']
        
        # Check rows.
        for row in board:
            if areDuplicates(excludeEmpty(row)):
                return False
        
        # Check columns.
        for j in range(9):
            column = [board[i][j] for i in range(9)]
            if areDuplicates(excludeEmpty(column)):
                return False
        
        # Check 3x3 sub-boxes.
        for rstart in range(0, 9, 3):
            for cstart in range(0, 9, 3):
                box = [board[rstart+i][cstart+j] for i in range(0,3) for j in range(0,3)]
                if areDuplicates(excludeEmpty(box)):
                    return False
        
        return True

File: array/valid-sudoku.js
// https://leetcode.com/problems/valid-sudoku/
// tags - array
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  const n = 9;
  const rows = new Array(n).fill(0).map(() => new Set());
  const cols = new Array(n).fill(0).map(() => new Set());
  const boxes = new Array(n).fill(0).map(() => new Set());
  for (let r = 0; r < n; r++) {
    for (let c = 0; c < n; c++) {
      const val = board[r][c];
      if (val === ".") continue;
      const box = Math.floor(r / 3) * 3 + Math.floor(c / 3);
      if (rows[r].has(val) || cols[c].has(val) || boxes[box].has(val)) {
        return false;
      }
      rows[r].add(val);
      cols[c].add(val);
      boxes[box].add(val);
    }
  }
  return true;
};
